File: decentralizedmessaging.java
import io.ipfs.api.IPFS;
import io.ipfs.api.NamedStreamable;
import io.ipfs.multihash.Multihash;
import io.libp2p.core.PeerId;
import io.libp2p.core.Stream;
import io.libp2p.core.multistream.Multistream;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;

public class Messaging {
    public static void main(String[] args) throws Exception {
        // Set up IPFS with libp2p as the transport
        IPFS ipfs = new IPFS("/ip4/127.0.0.1/tcp/5001");

        // Parse the libp2p address of the peer we want to connect to
        PeerId peerId = PeerId.createFromB58String("QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ");

        // Connect to the peer and set up a stream
        CompletableFuture<Stream> streamFuture = ipfs.swarm().connect(peerId);
        Stream stream = streamFuture.get();

        // Select the '/chat/1.0.0' protocol on the stream
        Multistream multistream = new Multistream(stream);
        multistream.handle("/chat/1.0.0", (protocol) -> {
            // Send a message to the peer
            protocol.write("Hello, world!".getBytes());

            // Read and print the response from the peer
            byte[] response = protocol.read();
            System.out.println(new String(response));

            return CompletableFuture.completedFuture(null);
        });
        multistream.select("/chat/1.0.0").get();

        // Close the stream
        stream.close().get();
    }
}

File: decentralizedmessaging.rs
extern crate libp2p;
extern crate futures;

use std::error::Error;
use libp2p::{
    Transport,
    Muxer,
    Security,
};
use libp2p::tcp::TcpConfig;
use libp2p::mplex::MplexConfig;
use libp2p::noise::NoiseConfig;
use libp2p::core::upgrade;
use libp2p::core::upgrade::{Negotiated, ReadOne};
use futures::prelude::*;

fn main() -> Result<(), Box<dyn Error>> {
    // Set up libp2p with the TCP and Mplex transports and the Noise encryption protocol
    let transport = TcpConfig::new().and_then(MplexConfig::new).and_then(NoiseConfig::new);
    let mut node = libp2p::build_development_node(transport)?;

    // Set up a listener for incoming messages
    node.set_stream_handler("/chat/1.0.0", |stream| {
        // Print the incoming message
        let message = stream.read_one().wait().unwrap();
        println!("{}", String::from_utf8(message).unwrap());

        // Close the stream
        stream.close().wait().unwrap();
        Ok(())
    });

    // Parse the libp2p address of the peer we want to connect to
    let peer_addr = "/ip4/104.131.131.82/tcp/4001/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ".parse()?;
    let peer_info = libp2p::PeerId::from_multiaddr(&peer_addr)?;

    // Connect to the peer
    let mut stream = node.new_stream().unwrap();
    stream.connect(peer_info).wait()?;

    // Send a message to the peer
    let message = b"Hello, world!";
    stream.write_all(message).wait()?;

    // Wait indefinitely
    loop {}
}
